Snowflake.Data/Core/Converter/JsonToStructuredTypeConverter.cs

```diff
@@ -0,0 +1,423 @@
+
+namespace Snowflake.Data.Core.Converter
+{
+    internal static class JsonToStructuredTypeConverter
+    {
+        private static readonly TimeConverter s_timeConverter = new TimeConverter();
+
+        public static T ConvertObject<T>(string sourceTypeName, List<FieldMetadata> fields, JObject value)
+        {
+            var type = typeof(T);
+            if (SFDataType.OBJECT.ToString().Equals(sourceTypeName, StringComparison.OrdinalIgnoreCase))
+            {
+                return (T) ConvertToObject(type, fields, value);
+            }
+
+            throw new Exception("OBJECT type was expected");
+        }
+
+        public static T[] ConvertArray<T>(string sourceTypeName, List<FieldMetadata> fields, JArray value)
+        {
+            var type = typeof(T[]);
+            var elementType = typeof(T);
+            if (SFDataType.ARRAY.ToString().Equals(sourceTypeName, StringComparison.OrdinalIgnoreCase))
+            {
+                return (T[]) ConvertToArray(type, elementType, fields, value);
+            }
+
+            throw new Exception("ARRAY type was expected");
+        }
+
+        public static Dictionary<TKey, TValue> ConvertMap<TKey, TValue>(string sourceTypeName, List<FieldMetadata> fields, JObject value)
+        {
+            var type = typeof(Dictionary<TKey, TValue>);
+            var keyType = typeof(TKey);
+            var valueType = typeof(TValue);
+            if (SFDataType.MAP.ToString().Equals(sourceTypeName, StringComparison.OrdinalIgnoreCase))
+            {
+                return (Dictionary<TKey, TValue>) ConvertToMap(type, keyType, valueType, fields, value);
+            }
+
+            throw new Exception("MAP type was expected");
+        }
+
+
+        private static object ConvertToObject(Type type, List<FieldMetadata> fields, JToken json)
+        {
+            var context = $"converting structured object to client type {type}";
+                if (json == null || json.Type == JTokenType.Null || json.Type == JTokenType.Undefined)
+                {
+                    return null;
+                }
+
+                if (json.Type != JTokenType.Object)
+                {
+                    throw new Exception("Json is not an object");
+                }
+
+                var jsonObject = (JObject)json;
+                var constructionMethod = GetConstructionMethod(type);
+                var objectBuilder = ObjectBuilderFactory.Create(type, fields?.Count ?? 0, constructionMethod);
+                context = $"converting structured object fields to client type {type}";
+                using (var metadataIterator = fields.GetEnumerator())
+                {
+                    using (var jsonEnumerator = jsonObject.GetEnumerator())
+                    {
+                        do
+                        {
+                            var nextMetadataAvailable = metadataIterator.MoveNext();
+                            var nextJsonAvailable = jsonEnumerator.MoveNext();
```

Comment:
We are sure that field names and field metadata are in the same order?

Reply:
why would that be? Do you have any information about it?